Add aliasscopes to computational kernels

vchuravy opened the pull request:
## Before  

```
[ Info: Test batched evaluation
  27.368 μs (0 allocations: 0 bytes)
  3.242 μs (0 allocations: 0 bytes)
  65.632 μs (5 allocations: 41.17 KiB)
Ae ≈ val = true
all(∂BBe .≈ ∂BB) = true
  58.104 μs (0 allocations: 0 bytes)
```

## After

```
[ Info: Test batched evaluation
  27.675 μs (0 allocations: 0 bytes)
  3.210 μs (0 allocations: 0 bytes)
  65.505 μs (5 allocations: 41.17 KiB)
Ae ≈ val = true
all(∂BBe .≈ ∂BB) = true
  42.560 μs (0 allocations: 0 bytes)
```
10x is still not great (that's the upper bound I expect).

cc: @wsmoses

wsmoses commented:
Would be good to see the printall! of, but on first glance my guess is that reduce is recursive and causes the recursive tape from: https://github.com/EnzymeAD/Enzyme.jl/pull/462

```
Note that this will cause a performance regression for codes with recursion. But it will no longer compile assertion.
We should fix this later to restore the better performance without the assertion occurring in GC cases.
```

vchuravy commented:
`reduce` is fully unrolled and inlined. This is a straight for loop in the llvm IR

wsmoses commented:
Yea in that case if you have the printall / printperf log that would be the thing to look at.

vchuravy commented:
```
origin call may need caching   %57 = call noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj({}** nonnull %current_task30.i, i64 noundef 8, {} addrspace(10)* noundef addrspacecast ({}* inttoptr (i64 139724971694544 to {}*) to {} addrspace(10)*)) #11, !dbg !142
Callsite   call void @ijl_throw({} addrspace(12)* %59) #12, !dbg !142 arg 0   %59 = addrspacecast {} addrspace(10)* %57 to {} addrspace(12)*, !dbg !142 uncacheable from origin   %57 = call noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj({}** nonnull %current_task30.i, i64 noundef 8, {} addrspace(10)* noundef addrspacecast ({}* inttoptr (i64 139724971694544 to {}*) to {} addrspace(10)*)) #11, !dbg !142
Load may need caching   %17 = load i64, i64 addrspace(11)* %16, align 8, !dbg !40, !tbaa !55, !range !22 due to   call void @llvm.memset.p13i8.i64(i8 addrspace(13)* nonnull align 8 %60, i8 noundef 0, i64 %61, i1 noundef false) #10, !dbg !126
Load may need caching   %.sroa.8.0.us186 = load i64, i64 addrspace(13)* %scevgep2, align 8, !dbg !68, !tbaa !69 due to   store double %52, double addrspace(13)* %54, align 8, !dbg !126, !tbaa !69, !noalias !87
Load may need caching   %.sroa.675.0.us187 = load i64, i64 addrspace(13)* %scevgep1, align 8, !dbg !68, !tbaa !69 due to   store double %52, double addrspace(13)* %54, align 8, !dbg !126, !tbaa !69, !noalias !87
Load may need caching   %.sroa.072.0.us188 = load i64, i64 addrspace(13)* %scevgep, align 8, !dbg !68, !tbaa !69 due to   store double %52, double addrspace(13)* %54, align 8, !dbg !126, !tbaa !69, !noalias !87
Caching instruction   %42 = mul i64 %8, %41 legalRecompute: 1 shouldRecompute: 0 tryLegalRecomputeCheck: 1
Caching instruction   %43 = mul i64 %8, %42 legalRecompute: 1 shouldRecompute: 0 tryLegalRecomputeCheck: 1
Caching instruction   %43 = mul i64 %8, %42 legalRecompute: 1 shouldRecompute: 0 tryLegalRecomputeCheck: 1
[ Info: Test batched evaluation
  26.854 μs (0 allocations: 0 bytes)
  2.794 μs (0 allocations: 0 bytes)
  73.384 μs (5 allocations: 41.17 KiB)
Ae ≈ val = true
all(∂BBe .≈ ∂BB) = true
  43.335 μs (0 allocations: 0 bytes)
```

## print_all

```
fter simplification :
; Function Attrs: mustprogress willreturn
define void @preprocess_julia_evalpool__2634_inner.1({} addrspace(10)* nocapture nonnull readonly align 16 dereferenceable(40) %0, [1 x {} addrspace(10)*] %1, [3 x {} addrspace(10)*] %2) local_unnamed_addr #9 !dbg !148 {
L10.i.lr.ph:
  %.fca.0.extract1 = extractvalue [1 x {} addrspace(10)*] %1, 0, !dbg !149
  %.fca.0.extract = extractvalue [3 x {} addrspace(10)*] %2, 0, !dbg !149
  %.fca.1.extract = extractvalue [3 x {} addrspace(10)*] %2, 1, !dbg !149
  %.fca.2.extract = extractvalue [3 x {} addrspace(10)*] %2, 2, !dbg !149
  %3 = call {}*** @julia.get_pgcstack() #10
  call void @llvm.assume(i1 noundef true) #10
  %4 = bitcast {} addrspace(10)* %.fca.0.extract to {} addrspace(10)* addrspace(10)*, !dbg !150
  %5 = getelementptr inbounds {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %4, i64 3, !dbg !153
  %6 = bitcast {} addrspace(10)* addrspace(10)* %5 to i64 addrspace(10)*, !dbg !153
  %7 = addrspacecast i64 addrspace(10)* %6 to i64 addrspace(11)*, !dbg !153
  %8 = load i64, i64 addrspace(11)* %7, align 8, !dbg !153, !tbaa !18, !range !22, !invariant.load !4
  call void @llvm.assume(i1 noundef true) #10
  %9 = bitcast {} addrspace(10)* %.fca.1.extract to {} addrspace(10)* addrspace(10)*, !dbg !154
  %10 = getelementptr inbounds {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %9, i64 3, !dbg !159
  %11 = bitcast {} addrspace(10)* addrspace(10)* %10 to i64 addrspace(10)*, !dbg !159
  %12 = addrspacecast i64 addrspace(10)* %11 to i64 addrspace(11)*, !dbg !159
  %13 = load i64, i64 addrspace(11)* %12, align 8, !dbg !159, !tbaa !18, !range !22, !invariant.load !4
  %.not = icmp eq i64 %13, %8, !dbg !162
  br i1 %.not, label %L10.i.1, label %L158.i, !dbg !163

L33.i:                                            ; preds = %L10.i.1
  call void @llvm.assume(i1 noundef true) #10
  %14 = bitcast {} addrspace(10)* %.fca.0.extract1 to { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(10)*, !dbg !164
  %15 = addrspacecast { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(10)* %14 to { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)*, !dbg !164
  %16 = getelementptr inbounds { i8 addrspace(13)*, i64, i16, i16, i32 }, { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)* %15, i64 0, i32 1, !dbg !164
  %17 = load i64, i64 addrspace(11)* %16, align 8, !dbg !164, !tbaa !55, !range !22
  %.not133 = icmp eq i64 %17, 0, !dbg !172
  br i1 %.not133, label %julia_evalpool__2634_inner.exit, label %L52.i, !dbg !174

L52.i:                                            ; preds = %L33.i
  %18 = bitcast {} addrspace(10)* %.fca.0.extract1 to [3 x i64] addrspace(13)* addrspace(10)*, !dbg !175
  %19 = addrspacecast [3 x i64] addrspace(13)* addrspace(10)* %18 to [3 x i64] addrspace(13)* addrspace(11)*, !dbg !175
  %20 = load [3 x i64] addrspace(13)*, [3 x i64] addrspace(13)* addrspace(11)* %19, align 8, !dbg !175, !tbaa !65, !nonnull !4
  %.not139 = icmp eq i64 %8, 0
  %21 = bitcast {} addrspace(10)* %.fca.0.extract to double addrspace(13)* addrspace(10)*
  %22 = addrspacecast double addrspace(13)* addrspace(10)* %21 to double addrspace(13)* addrspace(11)*
  %23 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %22, align 8
  %24 = bitcast {} addrspace(10)* %.fca.1.extract to double addrspace(13)* addrspace(10)*
  %25 = addrspacecast double addrspace(13)* addrspace(10)* %24 to double addrspace(13)* addrspace(11)*
  %26 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %25, align 8
  %27 = bitcast {} addrspace(10)* %.fca.2.extract to double addrspace(13)* addrspace(10)*
  %28 = addrspacecast double addrspace(13)* addrspace(10)* %27 to double addrspace(13)* addrspace(11)*
  %29 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %28, align 8
  %30 = bitcast {} addrspace(10)* %0 to double addrspace(13)* addrspace(10)*
  %31 = addrspacecast double addrspace(13)* addrspace(10)* %30 to double addrspace(13)* addrspace(11)*
  %32 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %31, align 16, !tbaa !65, !nonnull !4
  br i1 %.not139, label %julia_evalpool__2634_inner.exit.loopexit, label %L52.i.split.split.us, !dbg !176

L52.i.split.split.us:                             ; preds = %L52.i
  %.elt136 = getelementptr inbounds [3 x i64], [3 x i64] addrspace(13)* %20, i64 0, i64 2, !dbg !175
  %.elt134 = getelementptr inbounds [3 x i64], [3 x i64] addrspace(13)* %20, i64 0, i64 1, !dbg !175
  %.elt = getelementptr inbounds [3 x i64], [3 x i64] addrspace(13)* %20, i64 0, i64 0, !dbg !175
  %33 = add nuw nsw i64 %17, 1, !dbg !176
  br label %L90.i.lr.ph.us190, !dbg !176

L90.i.lr.ph.us190:                                ; preds = %L148.i.us202, %L52.i.split.split.us
  %iv = phi i64 [ %iv.next, %L148.i.us202 ], [ 0, %L52.i.split.split.us ]
  %34 = add i64 %iv, 2, !dbg !177
  %iv.next = add nuw nsw i64 %iv, 1, !dbg !177
  %scevgep2 = getelementptr [3 x i64], [3 x i64] addrspace(13)* %20, i64 %iv, i64 2, !dbg !177
  %scevgep1 = getelementptr [3 x i64], [3 x i64] addrspace(13)* %20, i64 %iv, i64 1, !dbg !177
  %scevgep = getelementptr [3 x i64], [3 x i64] addrspace(13)* %20, i64 %iv, i64 0, !dbg !177
  %.sroa.8.0.us186 = load i64, i64 addrspace(13)* %scevgep2, align 8, !dbg !177, !tbaa !69
  %.sroa.675.0.us187 = load i64, i64 addrspace(13)* %scevgep1, align 8, !dbg !177, !tbaa !69
  %.sroa.072.0.us188 = load i64, i64 addrspace(13)* %scevgep, align 8, !dbg !177, !tbaa !69
  %35 = add i64 %.sroa.072.0.us188, -1
  %36 = mul i64 %8, %35
  %37 = add i64 %.sroa.675.0.us187, -1
  %38 = mul i64 %8, %37
  %39 = add i64 %.sroa.8.0.us186, -1
  %40 = mul i64 %8, %39
  br label %L90.i.us191, !dbg !178

L90.i.us191:                                      ; preds = %L90.i.us191, %L90.i.lr.ph.us190
  %iv3 = phi i64 [ %iv.next4, %L90.i.us191 ], [ 0, %L90.i.lr.ph.us190 ]
  %value_phi18.i175.us193 = phi double [ 0.000000e+00, %L90.i.lr.ph.us190 ], [ %52, %L90.i.us191 ]
  %iv.next4 = add nuw nsw i64 %iv3, 1, !dbg !180
  %41 = add i64 %36, %iv3, !dbg !180
  %42 = getelementptr inbounds double, double addrspace(13)* %23, i64 %41, !dbg !180
  %43 = load double, double addrspace(13)* %42, align 8, !dbg !180, !tbaa !69, !alias.scope !87, !llvm.mem.parallel_loop_access !186
  %44 = add i64 %38, %iv3, !dbg !180
  %45 = getelementptr inbounds double, double addrspace(13)* %26, i64 %44, !dbg !180
  %46 = load double, double addrspace(13)* %45, align 8, !dbg !180, !tbaa !69, !alias.scope !87, !llvm.mem.parallel_loop_access !186
  %47 = add i64 %40, %iv3, !dbg !180
  %48 = getelementptr inbounds double, double addrspace(13)* %29, i64 %47, !dbg !180
  %49 = load double, double addrspace(13)* %48, align 8, !dbg !180, !tbaa !69, !alias.scope !87, !llvm.mem.parallel_loop_access !186
  %50 = fmul fast double %46, %43, !dbg !188
  %51 = fmul fast double %50, %49, !dbg !200
  %52 = fadd fast double %value_phi18.i175.us193, %51, !dbg !203
  %exitcond209.not = icmp eq i64 %iv.next4, %8, !dbg !204
  br i1 %exitcond209.not, label %L121.i.us199, label %L90.i.us191, !dbg !178, !llvm.loop !187

L121.i.us199:                                     ; preds = %L90.i.us191
  %53 = add nsw i64 %iv.next, -1, !dbg !206
  %54 = getelementptr inbounds double, double addrspace(13)* %32, i64 %53, !dbg !206
  store double %52, double addrspace(13)* %54, align 8, !dbg !206, !tbaa !69, !noalias !87
  %55 = add nsw i64 %34, -1, !dbg !208
  %exitcond211.not = icmp eq i64 %34, %33, !dbg !213
  br i1 %exitcond211.not, label %julia_evalpool__2634_inner.exit.loopexit5, label %L148.i.us202, !dbg !215

L148.i.us202:                                     ; preds = %L121.i.us199
  %.elt142.us203 = getelementptr inbounds [3 x i64], [3 x i64] addrspace(13)* %20, i64 %55, i64 0, !dbg !216
  %.elt144.us204 = getelementptr inbounds [3 x i64], [3 x i64] addrspace(13)* %20, i64 %55, i64 1, !dbg !216
  %.elt146.us205 = getelementptr inbounds [3 x i64], [3 x i64] addrspace(13)* %20, i64 %55, i64 2, !dbg !216
  %56 = add nuw i64 %34, 1, !dbg !217
  br label %L90.i.lr.ph.us190

L158.i:                                           ; preds = %L10.i.1, %L10.i.lr.ph
  %current_task30.i148 = getelementptr inbounds {}**, {}*** %3, i64 -12, !dbg !219
  %current_task30.i = bitcast {}*** %current_task30.i148 to {}**, !dbg !219
  %57 = call noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj({}** nonnull %current_task30.i, i64 noundef 8, {} addrspace(10)* noundef addrspacecast ({}* inttoptr (i64 140714559003088 to {}*) to {} addrspace(10)*)) #11, !dbg !219
  %58 = bitcast {} addrspace(10)* %57 to {} addrspace(10)* addrspace(10)*, !dbg !219
  store {} addrspace(10)* addrspacecast ({}* inttoptr (i64 140713983704128 to {}*) to {} addrspace(10)*), {} addrspace(10)* addrspace(10)* %58, align 8, !dbg !219, !tbaa !145
  %59 = addrspacecast {} addrspace(10)* %57 to {} addrspace(12)*, !dbg !219
  call void @ijl_throw({} addrspace(12)* %59) #12, !dbg !219
  unreachable, !dbg !219

julia_evalpool__2634_inner.exit.loopexit:         ; preds = %L52.i
  %60 = bitcast double addrspace(13)* %32 to i8 addrspace(13)*, !dbg !176
  %61 = shl nuw i64 %17, 3, !dbg !176
  call void @llvm.memset.p13i8.i64(i8 addrspace(13)* nonnull align 8 %60, i8 noundef 0, i64 %61, i1 noundef false) #10, !dbg !206
  br label %julia_evalpool__2634_inner.exit, !dbg !149

julia_evalpool__2634_inner.exit.loopexit5:        ; preds = %L121.i.us199
  br label %julia_evalpool__2634_inner.exit, !dbg !149

julia_evalpool__2634_inner.exit:                  ; preds = %julia_evalpool__2634_inner.exit.loopexit5, %julia_evalpool__2634_inner.exit.loopexit, %L33.i
  ret void, !dbg !149

L10.i.1:                                          ; preds = %L10.i.lr.ph
  call void @llvm.assume(i1 noundef true) #10
  %62 = bitcast {} addrspace(10)* %.fca.2.extract to {} addrspace(10)* addrspace(10)*, !dbg !154
  %63 = getelementptr inbounds {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %62, i64 3, !dbg !159
  %64 = bitcast {} addrspace(10)* addrspace(10)* %63 to i64 addrspace(10)*, !dbg !159
  %65 = addrspacecast i64 addrspace(10)* %64 to i64 addrspace(11)*, !dbg !159
  %66 = load i64, i64 addrspace(11)* %65, align 8, !dbg !159, !tbaa !18, !range !22, !invariant.load !4
  %.not.1 = icmp eq i64 %66, %8, !dbg !162
  br i1 %.not.1, label %L33.i, label %L158.i, !dbg !163
}

; Function Attrs: mustprogress willreturn
define internal void @diffejulia_evalpool__2634_inner.1({} addrspace(10)* nocapture nonnull readonly align 16 dereferenceable(40) %0, {} addrspace(10)* nocapture %"'", [1 x {} addrspace(10)*] %1, [3 x {} addrspace(10)*] %2, [3 x {} addrspace(10)*] %"'1") local_unnamed_addr #9 !dbg !220 {
L10.i.lr.ph:
  %"iv'ac" = alloca i64, align 8
  %"iv3'ac" = alloca i64, align 8
  %"'de" = alloca double, align 8
  store double 0.000000e+00, double* %"'de", align 8
  %"value_phi18.i175.us193'de" = alloca double, align 8
  store double 0.000000e+00, double* %"value_phi18.i175.us193'de", align 8
  %"'de16" = alloca double, align 8
  store double 0.000000e+00, double* %"'de16", align 8
  %_cache = alloca i64*, align 8
  %_cache26 = alloca i64*, align 8
  %_cache35 = alloca i64*, align 8
  %"'de42" = alloca double, align 8
  store double 0.000000e+00, double* %"'de42", align 8
  %"'de43" = alloca double, align 8
  store double 0.000000e+00, double* %"'de43", align 8
  %"'de46" = alloca double, align 8
  store double 0.000000e+00, double* %"'de46", align 8
  %"'de47" = alloca double, align 8
  store double 0.000000e+00, double* %"'de47", align 8
  %.fca.0.extract1 = extractvalue [1 x {} addrspace(10)*] %1, 0, !dbg !221
  %".fca.0.extract'ipev" = extractvalue [3 x {} addrspace(10)*] %"'1", 0, !dbg !221
  %.fca.0.extract = extractvalue [3 x {} addrspace(10)*] %2, 0, !dbg !221
  %".fca.1.extract'ipev" = extractvalue [3 x {} addrspace(10)*] %"'1", 1, !dbg !221
  %.fca.1.extract = extractvalue [3 x {} addrspace(10)*] %2, 1, !dbg !221
  %".fca.2.extract'ipev" = extractvalue [3 x {} addrspace(10)*] %"'1", 2, !dbg !221
  %.fca.2.extract = extractvalue [3 x {} addrspace(10)*] %2, 2, !dbg !221
  %3 = call {}*** @julia.get_pgcstack() #10
  call void @llvm.assume(i1 noundef true) #10
  %4 = bitcast {} addrspace(10)* %.fca.0.extract to {} addrspace(10)* addrspace(10)*, !dbg !222
  %5 = getelementptr inbounds {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %4, i64 3, !dbg !225
  %6 = bitcast {} addrspace(10)* addrspace(10)* %5 to i64 addrspace(10)*, !dbg !225
  %7 = addrspacecast i64 addrspace(10)* %6 to i64 addrspace(11)*, !dbg !225
  %8 = load i64, i64 addrspace(11)* %7, align 8, !dbg !225, !tbaa !18, !range !22, !invariant.load !4
  call void @llvm.assume(i1 noundef true) #10
  %9 = bitcast {} addrspace(10)* %.fca.1.extract to {} addrspace(10)* addrspace(10)*, !dbg !226
  %10 = getelementptr inbounds {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %9, i64 3, !dbg !231
  %11 = bitcast {} addrspace(10)* addrspace(10)* %10 to i64 addrspace(10)*, !dbg !231
  %12 = addrspacecast i64 addrspace(10)* %11 to i64 addrspace(11)*, !dbg !231
  %13 = load i64, i64 addrspace(11)* %12, align 8, !dbg !231, !tbaa !18, !range !22, !invariant.load !4
  %.not = icmp eq i64 %13, %8, !dbg !234
  br i1 %.not, label %L10.i.1, label %L158.i, !dbg !235

L33.i:                                            ; preds = %L10.i.1
  call void @llvm.assume(i1 noundef true) #10
  %14 = bitcast {} addrspace(10)* %.fca.0.extract1 to { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(10)*, !dbg !236
  %15 = addrspacecast { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(10)* %14 to { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)*, !dbg !236
  %16 = getelementptr inbounds { i8 addrspace(13)*, i64, i16, i16, i32 }, { i8 addrspace(13)*, i64, i16, i16, i32 } addrspace(11)* %15, i64 0, i32 1, !dbg !236
  %17 = load i64, i64 addrspace(11)* %16, align 8, !dbg !236, !tbaa !55, !range !22
  %.not133 = icmp eq i64 %17, 0, !dbg !244
  br i1 %.not133, label %julia_evalpool__2634_inner.exit, label %L52.i, !dbg !246

L52.i:                                            ; preds = %L33.i
  %18 = bitcast {} addrspace(10)* %.fca.0.extract1 to [3 x i64] addrspace(13)* addrspace(10)*, !dbg !247
  %19 = addrspacecast [3 x i64] addrspace(13)* addrspace(10)* %18 to [3 x i64] addrspace(13)* addrspace(11)*, !dbg !247
  %20 = load [3 x i64] addrspace(13)*, [3 x i64] addrspace(13)* addrspace(11)* %19, align 8, !dbg !247, !tbaa !65, !nonnull !4
  %.not139 = icmp eq i64 %8, 0
  %"'ipc13" = bitcast {} addrspace(10)* %".fca.0.extract'ipev" to double addrspace(13)* addrspace(10)*
  %21 = bitcast {} addrspace(10)* %.fca.0.extract to double addrspace(13)* addrspace(10)*
  %"'ipc14" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc13" to double addrspace(13)* addrspace(11)*
  %22 = addrspacecast double addrspace(13)* addrspace(10)* %21 to double addrspace(13)* addrspace(11)*
  %"'ipl15" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc14", align 8, !invariant.group !248
  %23 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %22, align 8, !invariant.group !249
  %"'ipc10" = bitcast {} addrspace(10)* %".fca.1.extract'ipev" to double addrspace(13)* addrspace(10)*
  %24 = bitcast {} addrspace(10)* %.fca.1.extract to double addrspace(13)* addrspace(10)*
  %"'ipc11" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc10" to double addrspace(13)* addrspace(11)*
  %25 = addrspacecast double addrspace(13)* addrspace(10)* %24 to double addrspace(13)* addrspace(11)*
  %"'ipl12" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc11", align 8, !invariant.group !250
  %26 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %25, align 8, !invariant.group !251
  %"'ipc7" = bitcast {} addrspace(10)* %".fca.2.extract'ipev" to double addrspace(13)* addrspace(10)*
  %27 = bitcast {} addrspace(10)* %.fca.2.extract to double addrspace(13)* addrspace(10)*
  %"'ipc8" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc7" to double addrspace(13)* addrspace(11)*
  %28 = addrspacecast double addrspace(13)* addrspace(10)* %27 to double addrspace(13)* addrspace(11)*
  %"'ipl9" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc8", align 8, !invariant.group !252
  %29 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %28, align 8, !invariant.group !253
  %"'ipc" = bitcast {} addrspace(10)* %"'" to double addrspace(13)* addrspace(10)*
  %30 = bitcast {} addrspace(10)* %0 to double addrspace(13)* addrspace(10)*
  %"'ipc6" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc" to double addrspace(13)* addrspace(11)*
  %31 = addrspacecast double addrspace(13)* addrspace(10)* %30 to double addrspace(13)* addrspace(11)*
  %"'ipl" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc6", align 16, !tbaa !65, !nonnull !4, !invariant.group !254
  %32 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %31, align 16, !tbaa !65, !nonnull !4
  br i1 %.not139, label %julia_evalpool__2634_inner.exit.loopexit, label %L52.i.split.split.us, !dbg !255

L52.i.split.split.us:                             ; preds = %L52.i
  %33 = add nuw nsw i64 %17, 1, !dbg !255
  %34 = add nsw i64 %17, -1, !dbg !255
  %35 = add nsw i64 %8, -1, !dbg !255
  %36 = add nuw i64 %34, 1, !dbg !255
  %37 = mul nuw i64 %36, 8, !dbg !255
  %38 = call noalias nonnull i8* @malloc(i64 %37), !dbg !255
  call void @llvm.memset.p0i8.i64(i8* %38, i8 0, i64 %37, i1 false), !dbg !255
  %_malloccache = bitcast i8* %38 to i64*, !dbg !255
  store i64* %_malloccache, i64** %_cache, align 8, !dbg !255, !invariant.group !256
  %39 = mul nuw i64 %36, 8, !dbg !255
  %40 = call noalias nonnull i8* @malloc(i64 %39), !dbg !255
  call void @llvm.memset.p0i8.i64(i8* %40, i8 0, i64 %39, i1 false), !dbg !255
  %_malloccache27 = bitcast i8* %40 to i64*, !dbg !255
  store i64* %_malloccache27, i64** %_cache26, align 8, !dbg !255, !invariant.group !257
  %41 = mul nuw i64 %36, 8, !dbg !255
  %42 = call noalias nonnull i8* @malloc(i64 %41), !dbg !255
  call void @llvm.memset.p0i8.i64(i8* %42, i8 0, i64 %41, i1 false), !dbg !255
  %_malloccache36 = bitcast i8* %42 to i64*, !dbg !255
  store i64* %_malloccache36, i64** %_cache35, align 8, !dbg !255, !invariant.group !258
  br label %L90.i.lr.ph.us190, !dbg !255

L90.i.lr.ph.us190:                                ; preds = %L148.i.us202, %L52.i.split.split.us
  %iv = phi i64 [ %iv.next, %L148.i.us202 ], [ 0, %L52.i.split.split.us ]
  %iv.next = add nuw nsw i64 %iv, 1, !dbg !259
  %43 = add i64 %iv, 2, !dbg !259
  %scevgep2 = getelementptr [3 x i64], [3 x i64] addrspace(13)* %20, i64 %iv, i64 2, !dbg !259
  %scevgep1 = getelementptr [3 x i64], [3 x i64] addrspace(13)* %20, i64 %iv, i64 1, !dbg !259
  %scevgep = getelementptr [3 x i64], [3 x i64] addrspace(13)* %20, i64 %iv, i64 0, !dbg !259
  %.sroa.8.0.us186 = load i64, i64 addrspace(13)* %scevgep2, align 8, !dbg !259, !tbaa !69
  %.sroa.675.0.us187 = load i64, i64 addrspace(13)* %scevgep1, align 8, !dbg !259, !tbaa !69
  %.sroa.072.0.us188 = load i64, i64 addrspace(13)* %scevgep, align 8, !dbg !259, !tbaa !69
  %44 = add i64 %.sroa.072.0.us188, -1
  %45 = mul i64 %8, %44
  %46 = add i64 %.sroa.675.0.us187, -1
  %47 = mul i64 %8, %46
  %48 = add i64 %.sroa.8.0.us186, -1
  %49 = mul i64 %8, %48
  %50 = load i64*, i64** %_cache, align 8, !dbg !260, !dereferenceable !262, !invariant.group !256
  %51 = getelementptr inbounds i64, i64* %50, i64 %iv, !dbg !260
  store i64 %49, i64* %51, align 8, !dbg !260, !invariant.group !263
  %52 = load i64*, i64** %_cache35, align 8, !dbg !260, !dereferenceable !262, !invariant.group !258
  %53 = getelementptr inbounds i64, i64* %52, i64 %iv, !dbg !260
  store i64 %45, i64* %53, align 8, !dbg !260, !invariant.group !264
  %54 = load i64*, i64** %_cache26, align 8, !dbg !260, !dereferenceable !262, !invariant.group !257
  %55 = getelementptr inbounds i64, i64* %54, i64 %iv, !dbg !260
  store i64 %47, i64* %55, align 8, !dbg !260, !invariant.group !265
  br label %L90.i.us191, !dbg !260

L90.i.us191:                                      ; preds = %L90.i.us191, %L90.i.lr.ph.us190
  %iv3 = phi i64 [ %iv.next4, %L90.i.us191 ], [ 0, %L90.i.lr.ph.us190 ]
  %value_phi18.i175.us193 = phi double [ 0.000000e+00, %L90.i.lr.ph.us190 ], [ %67, %L90.i.us191 ]
  %iv.next4 = add nuw nsw i64 %iv3, 1, !dbg !266
  %56 = add i64 %45, %iv3, !dbg !266
  %"'ipg49" = getelementptr inbounds double, double addrspace(13)* %"'ipl15", i64 %56, !dbg !266
  %57 = getelementptr inbounds double, double addrspace(13)* %23, i64 %56, !dbg !266
  %58 = load double, double addrspace(13)* %57, align 8, !dbg !266, !tbaa !69, !alias.scope !87, !llvm.mem.parallel_loop_access !272, !invariant.group !274
  %59 = add i64 %47, %iv3, !dbg !266
  %"'ipg48" = getelementptr inbounds double, double addrspace(13)* %"'ipl12", i64 %59, !dbg !266
  %60 = getelementptr inbounds double, double addrspace(13)* %26, i64 %59, !dbg !266
  %61 = load double, double addrspace(13)* %60, align 8, !dbg !266, !tbaa !69, !alias.scope !87, !llvm.mem.parallel_loop_access !272, !invariant.group !275
  %62 = add i64 %49, %iv3, !dbg !266
  %"'ipg" = getelementptr inbounds double, double addrspace(13)* %"'ipl9", i64 %62, !dbg !266
  %63 = getelementptr inbounds double, double addrspace(13)* %29, i64 %62, !dbg !266
  %64 = load double, double addrspace(13)* %63, align 8, !dbg !266, !tbaa !69, !alias.scope !87, !llvm.mem.parallel_loop_access !272, !invariant.group !276
  %65 = fmul fast double %61, %58, !dbg !277
  %66 = fmul fast double %65, %64, !dbg !289
  %67 = fadd fast double %value_phi18.i175.us193, %66, !dbg !292
  %exitcond209.not = icmp eq i64 %iv.next4, %8, !dbg !293
  br i1 %exitcond209.not, label %L121.i.us199, label %L90.i.us191, !dbg !260, !llvm.loop !273

L121.i.us199:                                     ; preds = %L90.i.us191
  %68 = add nsw i64 %iv.next, -1, !dbg !295
  %"'ipg50" = getelementptr inbounds double, double addrspace(13)* %"'ipl", i64 %68, !dbg !295
  %69 = getelementptr inbounds double, double addrspace(13)* %32, i64 %68, !dbg !295
  store double %67, double addrspace(13)* %69, align 8, !dbg !295, !tbaa !69, !alias.scope !297, !noalias !300
  %exitcond211.not = icmp eq i64 %43, %33, !dbg !302
  br i1 %exitcond211.not, label %julia_evalpool__2634_inner.exit.loopexit5, label %L148.i.us202, !dbg !307

L148.i.us202:                                     ; preds = %L121.i.us199
  br label %L90.i.lr.ph.us190

L158.i:                                           ; preds = %L10.i.1, %L10.i.lr.ph
  %current_task30.i148 = getelementptr inbounds {}**, {}*** %3, i64 -12, !dbg !308
  %current_task30.i = bitcast {}*** %current_task30.i148 to {}**, !dbg !308
  %70 = call noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj({}** nonnull %current_task30.i, i64 noundef 8, {} addrspace(10)* noundef addrspacecast ({}* inttoptr (i64 140714559003088 to {}*) to {} addrspace(10)*)) #11, !dbg !308
  %71 = bitcast {} addrspace(10)* %70 to {} addrspace(10)* addrspace(10)*, !dbg !308
  store {} addrspace(10)* addrspacecast ({}* inttoptr (i64 140713983704128 to {}*) to {} addrspace(10)*), {} addrspace(10)* addrspace(10)* %71, align 8, !dbg !308, !tbaa !145
  %72 = addrspacecast {} addrspace(10)* %70 to {} addrspace(12)*, !dbg !308
  call void @ijl_throw({} addrspace(12)* %72) #12, !dbg !308
  unreachable

julia_evalpool__2634_inner.exit.loopexit:         ; preds = %L52.i
  %"'ipc54" = bitcast double addrspace(13)* %"'ipl" to i8 addrspace(13)*, !dbg !255
  %73 = bitcast double addrspace(13)* %32 to i8 addrspace(13)*, !dbg !255
  %74 = shl nuw i64 %17, 3, !dbg !255
  call void @llvm.memset.p13i8.i64(i8 addrspace(13)* nonnull align 8 %73, i8 noundef 0, i64 %74, i1 noundef false) #10, !dbg !295
  br label %julia_evalpool__2634_inner.exit, !dbg !221

julia_evalpool__2634_inner.exit.loopexit5:        ; preds = %L121.i.us199
  br label %julia_evalpool__2634_inner.exit, !dbg !221

julia_evalpool__2634_inner.exit:                  ; preds = %julia_evalpool__2634_inner.exit.loopexit5, %julia_evalpool__2634_inner.exit.loopexit, %L33.i
  br label %invertjulia_evalpool__2634_inner.exit, !dbg !221

L10.i.1:                                          ; preds = %L10.i.lr.ph
  call void @llvm.assume(i1 noundef true) #10
  %75 = bitcast {} addrspace(10)* %.fca.2.extract to {} addrspace(10)* addrspace(10)*, !dbg !226
  %76 = getelementptr inbounds {} addrspace(10)*, {} addrspace(10)* addrspace(10)* %75, i64 3, !dbg !231
  %77 = bitcast {} addrspace(10)* addrspace(10)* %76 to i64 addrspace(10)*, !dbg !231
  %78 = addrspacecast i64 addrspace(10)* %77 to i64 addrspace(11)*, !dbg !231
  %79 = load i64, i64 addrspace(11)* %78, align 8, !dbg !231, !tbaa !18, !range !22, !invariant.load !4
  %.not.1 = icmp eq i64 %79, %8, !dbg !234
  br i1 %.not.1, label %L33.i, label %L158.i, !dbg !235

invertL10.i.lr.ph:                                ; preds = %invertL10.i.1
  ret void

invertL33.i:                                      ; preds = %invertjulia_evalpool__2634_inner.exit, %invertL52.i
  br label %invertL10.i.1

invertL52.i:                                      ; preds = %invertjulia_evalpool__2634_inner.exit.loopexit, %invertL52.i.split.split.us
  br label %invertL33.i

invertL52.i.split.split.us:                       ; preds = %invertL90.i.lr.ph.us190
  %80 = load i64, i64* %"iv'ac", align 8
  %forfree = load i64*, i64** %_cache, align 8, !dereferenceable !262, !invariant.group !256
  %81 = bitcast i64* %forfree to i8*
  call void @free(i8* nonnull %81), !dbg !221
  %82 = load i64, i64* %"iv'ac", align 8
  %forfree28 = load i64*, i64** %_cache26, align 8, !dereferenceable !262, !invariant.group !257
  %83 = bitcast i64* %forfree28 to i8*
  call void @free(i8* nonnull %83), !dbg !221
  %84 = load i64, i64* %"iv'ac", align 8
  %forfree37 = load i64*, i64** %_cache35, align 8, !dereferenceable !262, !invariant.group !258
  %85 = bitcast i64* %forfree37 to i8*
  call void @free(i8* nonnull %85), !dbg !221
  br label %invertL52.i

invertL90.i.lr.ph.us190:                          ; preds = %invertL90.i.us191
  %86 = load i64, i64* %"iv'ac", align 8
  %87 = icmp eq i64 %86, 0
  %88 = xor i1 %87, true
  br i1 %87, label %invertL52.i.split.split.us, label %incinvertL90.i.lr.ph.us190

incinvertL90.i.lr.ph.us190:                       ; preds = %invertL90.i.lr.ph.us190
  %89 = load i64, i64* %"iv'ac", align 8
  %90 = add nsw i64 %89, -1
  store i64 %90, i64* %"iv'ac", align 8
  br label %invertL148.i.us202

invertL90.i.us191:                                ; preds = %mergeinvertL90.i.us191_L121.i.us199, %incinvertL90.i.us191
  %91 = load double, double* %"'de", align 8
  store double 0.000000e+00, double* %"'de", align 8
  %92 = load double, double* %"value_phi18.i175.us193'de", align 8
  %93 = fadd fast double %92, %91
  store double %93, double* %"value_phi18.i175.us193'de", align 8
  %94 = load double, double* %"'de16", align 8
  %95 = fadd fast double %94, %91
  store double %95, double* %"'de16", align 8
  %96 = load double, double* %"'de16", align 8
  %97 = load i64, i64* %"iv3'ac", align 8
  %98 = load i64, i64* %"iv'ac", align 8
  %_unwrap = bitcast {} addrspace(10)* %.fca.2.extract to double addrspace(13)* addrspace(10)*
  %_unwrap17 = addrspacecast double addrspace(13)* addrspace(10)* %_unwrap to double addrspace(13)* addrspace(11)*
  %_unwrap18 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %_unwrap17, align 8, !invariant.group !253
  %_unwrap19 = add nsw i64 %17, -1
  %99 = add nuw i64 %_unwrap19, 1
  %100 = load i64*, i64** %_cache, align 8, !dereferenceable !262, !invariant.group !256
  %101 = getelementptr inbounds i64, i64* %100, i64 %98
  %102 = load i64, i64* %101, align 8, !invariant.group !263
  %_unwrap20 = add i64 %102, %97
  %_unwrap21 = getelementptr inbounds double, double addrspace(13)* %_unwrap18, i64 %_unwrap20
  %_unwrap22 = load double, double addrspace(13)* %_unwrap21, align 8, !dbg !266, !tbaa !69, !invariant.group !276
  %m0diffe = fmul fast double %96, %_unwrap22
  %103 = load i64, i64* %"iv3'ac", align 8
  %104 = load i64, i64* %"iv'ac", align 8
  %_unwrap23 = bitcast {} addrspace(10)* %.fca.1.extract to double addrspace(13)* addrspace(10)*
  %_unwrap24 = addrspacecast double addrspace(13)* addrspace(10)* %_unwrap23 to double addrspace(13)* addrspace(11)*
  %_unwrap25 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %_unwrap24, align 8, !invariant.group !251
  %105 = add nuw i64 %_unwrap19, 1
  %106 = load i64*, i64** %_cache26, align 8, !dereferenceable !262, !invariant.group !257
  %107 = getelementptr inbounds i64, i64* %106, i64 %104
  %108 = load i64, i64* %107, align 8, !invariant.group !265
  %_unwrap29 = add i64 %108, %103
  %_unwrap30 = getelementptr inbounds double, double addrspace(13)* %_unwrap25, i64 %_unwrap29
  %_unwrap31 = load double, double addrspace(13)* %_unwrap30, align 8, !dbg !266, !tbaa !69, !invariant.group !275
  %_unwrap32 = bitcast {} addrspace(10)* %.fca.0.extract to double addrspace(13)* addrspace(10)*
  %_unwrap33 = addrspacecast double addrspace(13)* addrspace(10)* %_unwrap32 to double addrspace(13)* addrspace(11)*
  %_unwrap34 = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %_unwrap33, align 8, !invariant.group !249
  %109 = add nuw i64 %_unwrap19, 1
  %110 = load i64*, i64** %_cache35, align 8, !dereferenceable !262, !invariant.group !258
  %111 = getelementptr inbounds i64, i64* %110, i64 %104
  %112 = load i64, i64* %111, align 8, !invariant.group !264
  %_unwrap38 = add i64 %112, %103
  %_unwrap39 = getelementptr inbounds double, double addrspace(13)* %_unwrap34, i64 %_unwrap38
  %_unwrap40 = load double, double addrspace(13)* %_unwrap39, align 8, !dbg !266, !tbaa !69, !invariant.group !274
  %_unwrap41 = fmul fast double %_unwrap31, %_unwrap40
  %m1diffe = fmul fast double %96, %_unwrap41
  store double 0.000000e+00, double* %"'de16", align 8
  %113 = load double, double* %"'de42", align 8
  %114 = fadd fast double %113, %m0diffe
  store double %114, double* %"'de42", align 8
  %115 = load double, double* %"'de43", align 8
  %116 = fadd fast double %115, %m1diffe
  store double %116, double* %"'de43", align 8
  %117 = load double, double* %"'de42", align 8
  %m0diffe44 = fmul fast double %117, %_unwrap40
  %m1diffe45 = fmul fast double %117, %_unwrap31
  store double 0.000000e+00, double* %"'de42", align 8
  %118 = load double, double* %"'de46", align 8
  %119 = fadd fast double %118, %m0diffe44
  store double %119, double* %"'de46", align 8
  %120 = load double, double* %"'de47", align 8
  %121 = fadd fast double %120, %m1diffe45
  store double %121, double* %"'de47", align 8
  %122 = load double, double* %"'de43", align 8
  store double 0.000000e+00, double* %"'de43", align 8
  %123 = load i64, i64* %"iv3'ac", align 8
  %124 = load i64, i64* %"iv'ac", align 8
  %"'ipc7_unwrap" = bitcast {} addrspace(10)* %".fca.2.extract'ipev" to double addrspace(13)* addrspace(10)*
  %"'ipc8_unwrap" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc7_unwrap" to double addrspace(13)* addrspace(11)*
  %"'ipl9_unwrap" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc8_unwrap", align 8, !invariant.group !252
  %"'ipg_unwrap" = getelementptr inbounds double, double addrspace(13)* %"'ipl9_unwrap", i64 %_unwrap20
  %125 = load double, double addrspace(13)* %"'ipg_unwrap", align 8, !dbg !266, !tbaa !69, !alias.scope !300, !noalias !297
  %126 = fadd fast double %125, %122
  store double %126, double addrspace(13)* %"'ipg_unwrap", align 8, !dbg !266, !tbaa !69, !alias.scope !300, !noalias !297
  %127 = load double, double* %"'de46", align 8
  store double 0.000000e+00, double* %"'de46", align 8
  %128 = load i64, i64* %"iv3'ac", align 8
  %129 = load i64, i64* %"iv'ac", align 8
  %"'ipc10_unwrap" = bitcast {} addrspace(10)* %".fca.1.extract'ipev" to double addrspace(13)* addrspace(10)*
  %"'ipc11_unwrap" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc10_unwrap" to double addrspace(13)* addrspace(11)*
  %"'ipl12_unwrap" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc11_unwrap", align 8, !invariant.group !250
  %"'ipg48_unwrap" = getelementptr inbounds double, double addrspace(13)* %"'ipl12_unwrap", i64 %_unwrap29
  %130 = load double, double addrspace(13)* %"'ipg48_unwrap", align 8, !dbg !266, !tbaa !69, !alias.scope !300, !noalias !297
  %131 = fadd fast double %130, %127
  store double %131, double addrspace(13)* %"'ipg48_unwrap", align 8, !dbg !266, !tbaa !69, !alias.scope !300, !noalias !297
  %132 = load double, double* %"'de47", align 8
  store double 0.000000e+00, double* %"'de47", align 8
  %133 = load i64, i64* %"iv3'ac", align 8
  %134 = load i64, i64* %"iv'ac", align 8
  %"'ipc13_unwrap" = bitcast {} addrspace(10)* %".fca.0.extract'ipev" to double addrspace(13)* addrspace(10)*
  %"'ipc14_unwrap" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc13_unwrap" to double addrspace(13)* addrspace(11)*
  %"'ipl15_unwrap" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc14_unwrap", align 8, !invariant.group !248
  %"'ipg49_unwrap" = getelementptr inbounds double, double addrspace(13)* %"'ipl15_unwrap", i64 %_unwrap38
  %135 = load double, double addrspace(13)* %"'ipg49_unwrap", align 8, !dbg !266, !tbaa !69, !alias.scope !300, !noalias !297
  %136 = fadd fast double %135, %132
  store double %136, double addrspace(13)* %"'ipg49_unwrap", align 8, !dbg !266, !tbaa !69, !alias.scope !300, !noalias !297
  %137 = load double, double* %"value_phi18.i175.us193'de", align 8
  store double 0.000000e+00, double* %"value_phi18.i175.us193'de", align 8
  %138 = load i64, i64* %"iv3'ac", align 8
  %139 = icmp eq i64 %138, 0
  %140 = xor i1 %139, true
  %141 = select fast i1 %140, double %137, double 0.000000e+00
  %142 = load double, double* %"'de", align 8
  %143 = fadd fast double %142, %137
  %144 = select fast i1 %139, double %142, double %143
  store double %144, double* %"'de", align 8
  br i1 %139, label %invertL90.i.lr.ph.us190, label %incinvertL90.i.us191

incinvertL90.i.us191:                             ; preds = %invertL90.i.us191
  %145 = load i64, i64* %"iv3'ac", align 8
  %146 = add nsw i64 %145, -1
  store i64 %146, i64* %"iv3'ac", align 8
  br label %invertL90.i.us191

invertL121.i.us199:                               ; preds = %mergeinvertL90.i.lr.ph.us190_julia_evalpool__2634_inner.exit.loopexit5, %invertL148.i.us202
  %147 = load i64, i64* %"iv'ac", align 8
  %"'ipc_unwrap" = bitcast {} addrspace(10)* %"'" to double addrspace(13)* addrspace(10)*
  %"'ipc6_unwrap" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc_unwrap" to double addrspace(13)* addrspace(11)*
  %"'ipl_unwrap" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc6_unwrap", align 16, !tbaa !65, !nonnull !4, !invariant.group !254
  %iv.next_unwrap = add nuw nsw i64 %147, 1
  %_unwrap51 = add nsw i64 %iv.next_unwrap, -1
  %"'ipg50_unwrap" = getelementptr inbounds double, double addrspace(13)* %"'ipl_unwrap", i64 %_unwrap51
  %148 = load double, double addrspace(13)* %"'ipg50_unwrap", align 8
  store double 0.000000e+00, double addrspace(13)* %"'ipg50_unwrap", align 8, !dbg !295, !tbaa !69, !alias.scope !300, !noalias !309
  %149 = load double, double* %"'de", align 8
  %150 = fadd fast double %149, %148
  store double %150, double* %"'de", align 8
  %151 = load i64, i64* %"iv'ac", align 8
  %_unwrap52 = add nsw i64 %8, -1
  br label %mergeinvertL90.i.us191_L121.i.us199

mergeinvertL90.i.us191_L121.i.us199:              ; preds = %invertL121.i.us199
  store i64 %_unwrap52, i64* %"iv3'ac", align 8
  br label %invertL90.i.us191

invertL148.i.us202:                               ; preds = %incinvertL90.i.lr.ph.us190
  br label %invertL121.i.us199

invertjulia_evalpool__2634_inner.exit.loopexit:   ; preds = %staging
  %_unwrap55 = shl nuw i64 %17, 3, !dbg !295
  %"'ipc_unwrap56" = bitcast {} addrspace(10)* %"'" to double addrspace(13)* addrspace(10)*, !dbg !295
  %"'ipc6_unwrap57" = addrspacecast double addrspace(13)* addrspace(10)* %"'ipc_unwrap56" to double addrspace(13)* addrspace(11)*, !dbg !295
  %"'ipl_unwrap58" = load double addrspace(13)*, double addrspace(13)* addrspace(11)* %"'ipc6_unwrap57", align 16, !tbaa !65, !nonnull !4, !invariant.group !254
  %"'ipc54_unwrap" = bitcast double addrspace(13)* %"'ipl_unwrap58" to i8 addrspace(13)*, !dbg !295
  call void @llvm.memset.p13i8.i64(i8 addrspace(13)* nonnull align 8 %"'ipc54_unwrap", i8 noundef 0, i64 %_unwrap55, i1 noundef false) #10, !dbg !295
  br label %invertL52.i

invertjulia_evalpool__2634_inner.exit.loopexit5:  ; preds = %staging
  %_unwrap59 = add nsw i64 %17, -1
  br label %mergeinvertL90.i.lr.ph.us190_julia_evalpool__2634_inner.exit.loopexit5

mergeinvertL90.i.lr.ph.us190_julia_evalpool__2634_inner.exit.loopexit5: ; preds = %invertjulia_evalpool__2634_inner.exit.loopexit5
  store i64 %_unwrap59, i64* %"iv'ac", align 8
  br label %invertL121.i.us199

invertjulia_evalpool__2634_inner.exit:            ; preds = %julia_evalpool__2634_inner.exit
  %.not139_unwrap = icmp eq i64 %8, 0
  br i1 %.not133, label %invertL33.i, label %staging

invertL10.i.1:                                    ; preds = %invertL33.i
  br label %invertL10.i.lr.ph

staging:                                          ; preds = %invertjulia_evalpool__2634_inner.exit
  br i1 %.not139_unwrap, label %invertjulia_evalpool__2634_inner.exit.loopexit, label %invertjulia_evalpool__2634_inner.exit.loopexit5
}
```

wsmoses commented:
Looks like it caches three arrays of ints since it cannot prove that the int loaded isn't overwritten by storing that double:
```
Load may need caching   %.sroa.8.0.us186 = load i64, i64 addrspace(13)* %scevgep2, align 8, !dbg !68, !tbaa !69 due to   store double %52, double addrspace(13)* %54, align 8, !dbg !126, !tbaa !69, !noalias !87
Load may need caching   %.sroa.675.0.us187 = load i64, i64 addrspace(13)* %scevgep1, align 8, !dbg !68, !tbaa !69 due to   store double %52, double addrspace(13)* %54, align 8, !dbg !126, !tbaa !69, !noalias !87
Load may need caching   %.sroa.072.0.us188 = load i64, i64 addrspace(13)* %scevgep, align 8, !dbg !68, !tbaa !69 due to   store double %52, double addrspace(13)* %54, align 8, !dbg !126, !tbaa !69, !noalias !87
```

wsmoses commented:
It looks like those (potentially overwritten) loads probably come from `basis:PooledSparseProduct`

wsmoses commented:
What if you constify basis?

cortner commented:
should I label this PR `WIP`?

cortner commented:
So I hand-coded a special case to check what I could get in principle: 
```
[ Info: Test batched evaluation (inplace)
  1.329 μs (0 allocations: 0 bytes)

[ Info: timing in-place pullback
  3.026 μs (0 allocations: 0 bytes)
```
So it is absolutely possible (in principle) to get close to the optimal factor 2. How close can enzyme get?

vchuravy commented:
> Thank you for looking into this. should I label this PR WIP?

The changes in this PR are independent of Enzyme, so I think we could land it irrespective.

cortner commented:
ok for me. I tested it and profiled again and can merge if you say this is ready?

cortner commented:
merging latest `main` into this PR should fix CI.

cortner commented:
@vchuravy  --- sorry to ping you again here. I assume from your last message that this can be merged and will do it later today unless you tell me otherwise.
